File: DP/maximum_value_contiguous_subsequence.cpp
#include<iostream>

using namespace std;

int main() {
	int n;
	cout << "Enter the number of elements: ";
	cin >> n;
	int a[n];
	cout << "Enter the elements: " << endl;
	for(int i=0;i<n;i++) {
		cin >> a[i];
	}
	int m[n];
	memset(m,0,n);
	if(a[0]<0) {
		m[0]=0;
	} else {
		m[0]=a[0];
	}
	for(int i=1;i<n;i++) {
		if(m[i-1]+a[i]<0) {
			m[i]=0;
		} else {
			m[i]=m[i-1]+a[i];
		}
	}

	int max=0;
	for(int i=0;i<n;i++) {
		if(max<=m[i]) {
			max=m[i];
		}
	}
	cout << "Maximum Contagious Sum: " << max;
	
}

File: src/main/java/in/sg/dsa/dp/MaximumValueContiguousSubsequence.java
package in.sg.dsa.dp;

import java.util.Scanner;

public class MaximumValueContiguousSubsequence {

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    System.out.print("Enter the number of elements: ");
    int n = scanner.nextInt();
    int[] a = new int[n];
    System.out.println("Enter the elements: ");
    for (int i = 0; i < n; i++) {
      a[i] = scanner.nextInt();
    }

    if (n == 0) {
      System.out.println("Maximum Contiguous Sum: 0");
      scanner.close();
      return;
    }

    int maxSoFar = a[0];
    int maxEndingHere = a[0];

    for (int i = 1; i < n; i++) {
      maxEndingHere = Math.max(a[i], maxEndingHere + a[i]);
      maxSoFar = Math.max(maxSoFar, maxEndingHere);
    }

    System.out.println("Maximum Contiguous Sum: " + maxSoFar);
    scanner.close();
  }
}
